src/ripple/shamap/impl/SHAMapInnerNode.cpp

```diff
@@ -33,14 +31,90 @@
 #include <openssl/sha.h>
 
 #include <algorithm>
-#include <array>
 #include <iterator>
-#include <mutex>
 #include <utility>
 
+// This is used for the _mm_pause instruction:
+#include <immintrin.h>
+
 namespace ripple {
 
-std::mutex SHAMapInnerNode::childLock;
+/** A specialized 16-way spinlock used to protect inner node branches.
+
+    This class packs 16 separate spinlocks into a single 16-bit value. It makes
+    it possible to lock any one lock at once or, alternatively, all together.
+
+    The implementation tries to use portable constructs but has to be low-level
+    for performance.
+ */
+class SpinBitlock
+{
+private:
+    std::atomic<std::uint16_t>& bits_;
+    std::uint16_t mask_;
+
+public:
+    SpinBitlock(std::atomic<std::uint16_t>& lock)
+        : bits_(lock), mask_(0xFFFF)
+    {
+    }
+
+    SpinBitlock(std::atomic<std::uint16_t>& lock, int index)
+        : bits_(lock), mask_(1 << index)
+    {
+        assert(index >= 0 && index < 16);
+    }
+
+    [[nodiscard]]
+    bool
+    try_lock()
+    {
+        // If we want to grab all the individual bitlocks at once we cannot
+        // use `fetch_or`! To see why, imagine that `lock_ == 0x0020` which
+        // means that the `fetch_or` would return `0x0020` but all the bits
+        // would already be (incorrectly!) set. Oops!
+        std::uint16_t expected = 0;
+
+        if (mask_ != 0xFFFF)
+            return (bits_.fetch_or(mask_, std::memory_order_acquire) & mask_) ==
+                expected;
+
+        return bits_.compare_exchange_weak(
+            expected,
+            mask_,
+            std::memory_order_acquire,
+            std::memory_order_relaxed);
+    }
+
+    void
+    lock()
+    {
+        // Testing suggests that 99.9999% of the time this will succeed, so
+        // we try to optimize the fast path.
+        if (try_lock()) [[likely]]
+            return;
+
+        do
+        {
+            // We try to spin for a few times:
+            for (int i = 0; i != 100; ++i)
+            {
+                if (try_lock())
+                    return;
+
+                _mm_pause();
```

Comment:
So `__mm_pause` is a valid intrinsic in Linux, Windows and MacOS and **SSE2** is part of the "base" AMD64 instruction set, which we are guaranteed to have since we _only_ compile in 64-bit mode.

I'm not overly concerned about this and your other proposed solution may eliminate the need for the lock altogether anyways, @seelabs.

Reply:
@nbougalis If we only care about intel, then I agree with you. I was thinking since we support mac for development, and newer macs use ARM, we should segregate code like this. But it's not that big a deal either. I'm fine keeping it as-is.

Reply:
Why not use `std::this_thread::yield()` instead?

Reply:
@HowardHinnant `__mm_pause` doesn't actually interact with the scheduler like `yield` would, and we don't want to yield right away. If we iterate too much we do `yield` below.

`__mm_pause` is really purpose built for spin-wait use-cases like this (see: https://www.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/intrinsics/intrinsics-for-sse2/miscellaneous-functions-and-intrinsics/pause-intrinsic.html?wapkw=_mm_pause).